### DISCUSSION TITLE
How to force System.Text.Json deserializer to throw an exception when property is missing in JSON?

### DISCUSSION THREAD

**Discussion by InternalHigh** [dotnet/runtime#83362] [General] [answered]
The new UnmappedMemberHandling doesn't seem to handle this direction: Type has property, but JSON does not provide property.
```
class Dto
{
    public string Name { get; set; }
}

var json = "{}";

var dto = JsonSerializer.Deserialize<Dto>(json, new JsonSerializerOptions() // does not throw
{
    UnmappedMemberHandling = JsonUnmappedMemberHandling.Disallow
});
```

I need *all* properties to be required.

**Comment by pinkfloydx33** [comment]
Doesn't `[JsonRequired]` throw on missing values?

**Comment by InternalHigh** [comment, marked as answer]
> Doesn't `[JsonRequired]` throw on missing values?

Yes but i don't want to add it to all properties. I want all properties to be required by default.

**Reply by eiriktsarpalis** [reply]
Assuming you need to do this you can use contract customization to mark every property as required:

https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/custom-contracts

**Reply by Xyncgas** [reply]
why can't we just somehow use these api by saying ```select``` properties on object then make sure they are ```required``` when deserialized, imagine these api can be used like a database, instead of nested options inside JsonOptions it's so noisy

**Reply by krwq** [reply]
@Xyncgas to simply respond it's because producing contract is relatively expensive while cache is cheap. You cannot edit something in the cache because options wouldn't be thread safe but you can edit before it's created and therefore you do have access to resolver but not the cache. Also your resolver code is like 5 lines of code with resolver vs 3 with hypothetical APIs.

All you really need is (untested):

```csharp
options.TypeInfoResolver = new DefaultJsonTypeInfoResolver()
{
  Modifiers =  {
  ti =>
    {
      foreach (var prop in ti.Properties)
      {
        prop.IsRequired = true;
      }
    }
  };
};
```
